src/WriterImpl.cpp

```diff
@@ -0,0 +1,970 @@
+               lineGroupProto.set("cartesianBounds", bbox);
+
+               StructureNode sbox = StructureNode(imf_);
+               sbox.set("rangeMinimum", FloatNode(imf_, 0., E57_SINGLE,
+                     data3DHeader.pointFields.pointRangeMinimum,	data3DHeader.pointFields.pointRangeMaximum));
+               sbox.set("rangeMaximum", FloatNode(imf_, 0., E57_SINGLE,
+                     data3DHeader.pointFields.pointRangeMinimum,	data3DHeader.pointFields.pointRangeMaximum));
+               sbox.set("elevationMinimum", FloatNode(imf_, 0., E57_SINGLE,
+                     data3DHeader.pointFields.angleMinimum, data3DHeader.pointFields.angleMaximum));
+               sbox.set("elevationMaximum", FloatNode(imf_, 0., E57_SINGLE,
+                     data3DHeader.pointFields.angleMinimum, data3DHeader.pointFields.angleMaximum));
+               sbox.set("azimuthStart", FloatNode(imf_, 0., E57_SINGLE,
+                     data3DHeader.pointFields.angleMinimum, data3DHeader.pointFields.angleMaximum));
+               sbox.set("azimuthEnd", FloatNode(imf_, 0., E57_SINGLE,
+                     data3DHeader.pointFields.angleMinimum, data3DHeader.pointFields.angleMaximum));
+               lineGroupProto.set("sphericalBounds", sbox);
+         */
+         // Make empty codecs vector for use in creating groups CompressedVector.
+         // If this vector is empty, it is assumed that all fields will use the BitPack codec.
+         VectorNode lineGroupCodecs = VectorNode( imf_, true );
+
+         // Create CompressedVector for storing groups.
+         // Path Name: "/data3D/0/pointGroupingSchemes/groupingByLine/groups".
+         // We use the prototype and empty codecs tree from above.
+         // The CompressedVector will be filled by code below.
+         CompressedVectorNode groups = CompressedVectorNode( imf_, lineGroupProto, lineGroupCodecs );
+         groupingByLine.set( "groups", groups );
+      }
+
+      // Make a prototype of datatypes that will be stored in points record.
+      // This prototype will be used in creating the points CompressedVector.
+      // Using this proto in a CompressedVector will define path names like:
+      //      "/data3D/0/points/0/cartesianX"
+      StructureNode proto = StructureNode( imf_ );
+
+      // Because ScaledInteger min/max are the raw integer min/max, we must calculate them from the data min/max
+      double pointRangeScale = data3DHeader.pointFields.pointRangeScaledInteger;
+      double pointRangeOffset = 0.;
+      int64_t pointRangeMinimum =
+         (int64_t)floor( ( data3DHeader.pointFields.pointRangeMinimum - pointRangeOffset ) / pointRangeScale + .5 );
+      int64_t pointRangeMaximum =
+         (int64_t)floor( ( data3DHeader.pointFields.pointRangeMaximum - pointRangeOffset ) / pointRangeScale + .5 );
+
+      if ( data3DHeader.pointFields.cartesianXField )
+      {
+         if ( data3DHeader.pointFields.pointRangeScaledInteger > 0. )
```

Comment:
Suggestion: make `pointRangeScale` const above (I think it's const?) and use it to replace `data3DHeader.pointFields.pointRangeScaledInteger` in the following if/elses to make it more readable.

(Same for `data3DHeader.pointFields.pointRangeMinimum` and `data3DHeader.pointFields.pointRangeMaximum`)

Reply:
I did this and bit of other improvements in this section as well - reuse of the proto classes etc.

Reply:
I hope it is safe to reuse the nodes like this.

Reply:
Actually, I have just found a problem with this. Let me fix it first.

Reply:
Looking at the code, I don't think it is a good idea to share the same node for multiple fields.

I have reworked the implementation, so now it works in a similar way like the old code, but still avoids most of the repetition.